### PR TITLE
Add top-k classification accuracy metrics

### PR DESCRIPTION
**Pull request by ramananbalakrishnan**
### Synopsis

Implement a metric to track top-k accuracy rate in classification problems.

### Details

Related to #3986 . 

Added an overall metric called `top_k_categorical_accuracy()` to calculate accuracy rates while allowing for any of the top k values from `y_pred` to match against `y_true`. I have tried to keep the output format for the new function as close to the existing `categorical_accuracy()` function as possible.

For TensorFlow, this feature already exists as  [`tf.nn.in_top_k()`](https://www.tensorflow.org/versions/r0.11/api_docs/python/nn.html#in_top_k). With some re-shaping to suit tf's expectation, the existing function is used as-is.

For Theano, I have made use of `theano.tensor.argsort` to achieve the same result.

A simple test function has also been added to demonstrate expected operation.

### CODE REVIEW

File: keras/backend/tensorflow_backend.py [hunk 1]
```diff
@@ -1486,6 +1486,15 @@ def l2_normalize(x, axis):
     return tf.nn.l2_normalize(x, dim=axis)
 
 
+def in_top_k(predictions, targets, k):
+    '''Given tensors representing predictions and targets, evaluates whether the
+    targets are within the top k predictions
+    '''
+    y_pred = reshape(predictions, [1, shape(predictions)[0]])
+    y_target = reshape(argmax(targets), [1])
+    return cast(any(tf.nn.in_top_k(y_pred, y_target, k)), 'float32')
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: `any` seems it doesn't belong here. You'll want this to return an array. Additionally, I would prefer not casting to `float32` and use the native TF `bool` type instead (i.e. this function should be a thin wrapper over `tf.nn.in_top_k`). The fact that Theano does not have a bool type has to be dealt with in the Theano backend and should not affect the TensorFlow backend.

File: keras/backend/theano_backend.py [hunk 2]
```diff
@@ -1043,6 +1043,14 @@ def l2_normalize(x, axis):
     return x / norm
 
 
+def in_top_k(predictions, targets, k):
+    '''Given tensors representing predictions and targets, evaluates whether the
+    targets are within the top k predictions
+    '''
+    top_k_pred = T.argsort(predictions)[-k:]
+    return cast(any(equal(top_k_pred, T.argmax(targets))), 'float32')
```
Comment (fchollet) [inline keras/backend/theano_backend.py]: It seems preferable to use `int` to encode bools in Theano, than float32.

File: tests/keras/test_metrics.py [hunk 3]
```diff
@@ -53,5 +53,15 @@ def test_sparse_metrics():
         assert K.eval(metric(y_a, y_b)).shape == ()
 
 
+def test_top_k_categorical_accuracy():
+    y_true = K.variable(np.array([0, 1, 0]))
+    y_pred = K.variable(np.array([0.3, 0.2, 0.1]))
```
Comment (fchollet) [inline tests/keras/test_metrics.py]: This should be tested on batches, not on single samples.

File: keras/backend/tensorflow_backend.py [hunk 4]
```diff
@@ -1486,6 +1486,15 @@ def l2_normalize(x, axis):
     return tf.nn.l2_normalize(x, dim=axis)
 
 
+def in_top_k(predictions, targets, k):
+    '''Given tensors representing predictions and targets, evaluates whether the
+    targets are within the top k predictions
+    '''
```
Comment (fchollet) [inline keras/backend/tensorflow_backend.py]: The docstring should specify that the inputs are supposed to be batches. It should also describe the output (shape, type and contents).

**Comment by ramananbalakrishnan** [comment]
@fchollet , thank you for the review.

I had completely missed the point that the function is supposed to work on batches of input and not individual vectors. Made the following changes, as requested in the review.

- [x] docstring - added more information to mirror existing information [`tf.nn.in_top_k`](https://www.tensorflow.org/versions/r0.11/api_docs/python/nn.html#in_top_k)
- [x] tensorflow_backend - simplified the code to make it a (very) thin wrapper
- [x] theano_backend - modified to work properly on batches and to keep using native output type (`int`) from `equal`
- [x] test_metrics - testing is now against batches of input and covers more comprehensive cases

**Comment by ramananbalakrishnan** [comment]
I have added a default of `k=5` given that top-5 rates seem to most common after top-1.

Otherwise, simply adding `'top_k_categorical_accuracy'` to `metrics[]` in `model.compile` would not work as [`training.py`](https://github.com/fchollet/keras/blob/4de7eaa6a80fd4257b866a6b695450c40b72dd28/keras/engine/training.py#L663) only passes `y_true` and `y_pred`.

If there is a planned path forward to allow passing additional arguments to metrics/objectives functions via `model.compile`, I can look into implementing that in a separate pull request.

_side note_: I guess a similar problem would also occur when wanting higher `beta` values in the newly introduced [`fbeta_score`](https://github.com/fchollet/keras/blob/4de7eaa6a80fd4257b866a6b695450c40b72dd28/keras/metrics.py#L96) metric.

**Comment by fchollet** [comment]
LGTM

**Comment by fchollet** [comment]
> If there is a planned path forward to allow passing additional arguments to metrics/objectives functions via model.compile, I can look into implementing that in a separate pull request.

This is not an issue. One may just use a closure:

```python
def top_10_acc(y_true, y_pred):
   return top_k_categorical_accuracy(y_true, y_pred, 10)
```

**Comment by ramananbalakrishnan** [comment]
got it. thanks for merging in.
